webapp/channels/src/components/admin_console/access_control/editors/table_editor/table_editor.tsx

```diff
@@ -89,204 +75,279 @@ const parseExpression = async (expr: string): Promise<TableRow[]> => {
+// The biggest limitation is that all expressions are ANDed together, so it's not possible to
+// have OR logic.
 function TableEditor({
     value,
     onChange,
     onValidate,
     disabled = false,
     userAttributes,
+    onParseError,
+    actions,
 }: TableEditorProps): JSX.Element {
     const {formatMessage} = useIntl();
+
     const [rows, setRows] = useState<TableRow[]>([]);
     const [showTestResults, setShowTestResults] = useState(false);
     const [showHelpModal, setShowHelpModal] = useState(false);
+    const [autoOpenAttributeMenuForRow, setAutoOpenAttributeMenuForRow] = useState<number | null>(null);
 
-    // Update rows when value changes externally
+    // Effect to parse the incoming CEL expression string (value prop)
+    // and update the internal rows state. Handles errors during parsing.
     useEffect(() => {
-        parseExpression(value).then((rows) => {
-            setRows(rows);
-        });
-    }, [value]);
-
-    // Update the CEL expression when table changes
-    const updateExpression = (newRows: TableRow[]) => {
-        const validRows = newRows.filter((row) => row.attribute && row.values.length > 0);
-        const expr = validRows.map((row) => {
-            if (row.operator === 'is') {
-                return `user.attributes.${row.attribute} == "${row.values[0]}"`;
+        actions.getVisualAST(value).then((result) => {
+            if (result.error) {
+                setRows([]);
+                onParseError(result.error.message);
+                return;
             }
 
-            if (row.operator === 'is not') {
-                return `user.attributes.${row.attribute} != "${row.values[0]}"`;
+            setRows(parseExpression(result.data));
+        }).catch((err) => {
+            setRows([]);
+            if (onValidate) {
+                onValidate(false);
             }
-
-            if (row.operator === 'starts with') {
-                return `user.attributes.${row.attribute}.startsWith("${row.values[0]}")`;
+            onParseError(err.message);
+        });
+    }, [value, onValidate, onParseError]);
+
+    // Converts the internal rows state back into a CEL expression string
+    // and calls the onChange and onValidate props.
+    const updateExpression = useCallback((newRows: TableRow[]) => {
+        const rowsThatCanFormExpressions = newRows.filter((row) => row.attribute); // Only include rows that have an attribute selected
+
+        const expr = rowsThatCanFormExpressions.map((row) => {
+            const attributeExpr = `user.attributes.${row.attribute}`;
+            const config = OPERATOR_CONFIG[row.operator];
+
+            if (!config) {
+                // Fallback for unknown operators, defaulting to 'in' logic
+                // This handles cases where row.operator might be an unexpected string.
+                const valuesStr = row.values.map((val: string) => `"${val}"`).join(', ');
```

Comment:
> Do we need to worry about escaping or unescaping these values at all?

https://github.com/mattermost/mattermost/pull/30785#discussion_r2089039885 comment from @hmhealey 

I don't think so, this will eventually error out during the save (server will try to parse expression into AST to do several checks and validations).